packages/in-app-notification-dispatcher/src/index.ts

```diff
@@ -0,0 +1,287 @@
+): Promise<void> {
+  match(decodedMessage)
+    .with(
+      {
+        type: P.union(
+          "NewPurposeVersionWaitingForApproval",
+          "PurposeWaitingForApproval",
+          "PurposeVersionRejected",
+          "PurposeVersionActivated",
+          "DraftPurposeDeleted",
+          "WaitingForApprovalPurposeDeleted",
+          "PurposeAdded",
+          "DraftPurposeUpdated",
+          "PurposeActivated",
+          "PurposeArchived",
+          "PurposeVersionOverQuotaUnsuspended",
+          "PurposeVersionSuspendedByConsumer",
+          "PurposeVersionSuspendedByProducer",
+          "PurposeVersionUnsuspendedByConsumer",
+          "PurposeVersionUnsuspendedByProducer",
+          "WaitingForApprovalPurposeVersionDeleted",
+          "NewPurposeVersionActivated",
+          "PurposeCloned",
+          "PurposeDeletedByRevokedDelegation",
+          "PurposeVersionArchivedByRevokedDelegation"
+        ),
+      },
+      () => {
+        logger.info(
+          `No need to send an in-app notification for ${decodedMessage.type} message`
+        );
+      }
+    )
+    .exhaustive();
+}
+
+export async function handleAgreementMessage(
+  decodedMessage: AgreementEventEnvelopeV2,
+  logger: Logger
+): Promise<void> {
+  match(decodedMessage)
+    .with(
+      {
+        type: P.union(
+          "AgreementActivated",
+          "AgreementSubmitted",
+          "AgreementRejected",
+          "AgreementAdded",
+          "AgreementDeleted",
+          "DraftAgreementUpdated",
+          "AgreementUnsuspendedByProducer",
+          "AgreementUnsuspendedByConsumer",
+          "AgreementUnsuspendedByPlatform",
+          "AgreementArchivedByConsumer",
+          "AgreementArchivedByUpgrade",
+          "AgreementUpgraded",
+          "AgreementSuspendedByProducer",
+          "AgreementSuspendedByConsumer",
+          "AgreementSuspendedByPlatform",
+          "AgreementConsumerDocumentAdded",
+          "AgreementConsumerDocumentRemoved",
+          "AgreementSetDraftByPlatform",
+          "AgreementSetMissingCertifiedAttributesByPlatform",
+          "AgreementDeletedByRevokedDelegation",
+          "AgreementArchivedByRevokedDelegation"
+        ),
+      },
+      () => {
+        logger.info(
+          `No need to send an in-app notification for ${decodedMessage.type} message`
+        );
+      }
+    )
+    .exhaustive();
+}
+
+function processMessage(topicHandlers: TopicHandlers) {
+  return async (messagePayload: EachMessagePayload): Promise<void> => {
+    const { catalogTopic, agreementTopic, purposeTopic } = topicHandlers;
+
+    const { decodedMessage, handleMessage } = match(messagePayload.topic)
+      .with(catalogTopic, () => {
+        const decodedMessage = decodeKafkaMessage(
+          messagePayload.message,
+          EServiceEventV2
+        );
+
+        const handleMessage = handleCatalogMessage.bind(null, decodedMessage);
+
+        return { decodedMessage, handleMessage };
+      })
```

Comment:
minor, but if these could be changed to just return `[EServiceEventV2, handleCatalogMessage]` and then the decoding was done afterwards we'd reduce this duplication